pcap-dpdk.c

```diff
@@ -0,0 +1,742 @@
+	}else{
+		RTE_LOG(INFO, USER1, "dpdk: lcore %u has nothing to do\n", lcore_id);
+	}
+	//only use master lcore
+	if (lcore_id != master_lcore_id){
+		return;
+	}
+	while( max_cnt==-1 || pkt_cnt < max_cnt){
+		if (pd->break_loop){
+			break;
+		}
+		nb_rx = (int)rte_eth_rx_burst(portid, 0, pkts_burst, MAX_PKT_BURST);
+		pkt_cnt += nb_rx;
+		for ( i = 0; i < nb_rx; i++) {
+			m = pkts_burst[i];
+			calculate_timestamp(&(pd->ts_helper),&(pcap_header.ts));
+			data_len = rte_pktmbuf_data_len(m);
+			pcap_header.caplen = data_len; 
+			pcap_header.len = data_len; 
+			// volatile prefetch
+			rte_prefetch0(rte_pktmbuf_mtod(m, void *));
+			bp = NULL;
+			if (m->nb_segs == 1)
+			{
+				bp = rte_pktmbuf_mtod(m, u_char *);
+			}else{
+				if (m->pkt_len <= ETHER_MAX_JUMBO_FRAME_LEN)
+				{
+					gather_len = dpdk_gather_data(pd->pcap_tmp_buf, m);
+					bp = pd->pcap_tmp_buf;
+					pcap_header.caplen = gather_len;
+					pcap_header.len = gather_len;
+				}else{
+					// size too large
+					// why only free this pkt
+					rte_pktmbuf_free(m);
+				}
+			}
+			if (bp){
+				//default accpet all
+				is_accepted=1;
+				if (pd->filter_in_userland && p->fcode.bf_insns!=NULL)
+				{
+					if (!pcap_filter(p->fcode.bf_insns, bp, pcap_header.len, pcap_header.caplen)){
+						//rejected
+						is_accepted=0;
+					}
+				}
+				if (is_accepted){
+					cb(cb_arg, &pcap_header, bp);
+				}else{
+					pd->bpf_drop++;
+				}
+			}
+		}
+	}	
+	pd->rx_pkts = pkt_cnt;
+}
+static int launch_one_lcore(void *dpdk_user)
+{
+	dpdk_dispatch_inter(dpdk_user);
+	return 0;
+}
+static int pcap_dpdk_dispatch(pcap_t *p, int max_cnt, pcap_handler cb, u_char *pcap_user)
+{
+	unsigned lcore_id = 0;	
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	pd->rx_pkts=0;
+	pd->cb = cb;
+	pd->cb_arg = pcap_user;
+	pd->max_cnt = max_cnt;
+	pd->orig = p;
+	void *dpdk_user = p;	
+	// launch_one_lcore func will be called on every lcore include master core.
+	rte_eal_mp_remote_launch(launch_one_lcore, dpdk_user, CALL_MASTER);
+	RTE_LCORE_FOREACH_SLAVE(lcore_id) {
+		if (rte_eal_wait_lcore(lcore_id) < 0) {
+			break;
+		}
+	}
+	return pd->rx_pkts;	
+}
+
+static int pcap_dpdk_inject(pcap_t *p, const void *buf _U_, int size _U_)
+{
+	//not implemented yet
+	pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+		errno, "dpdk error: Inject function has not be implemented yet");
+	return PCAP_ERROR;
+}
+
+static void pcap_dpdk_close(pcap_t *p)
+{
+	struct pcap_dpdk *pd = p->priv;
+	if (pd==NULL)
+	{
+		return;
+	} 
+	if (pd->must_clear_promisc)
+	{
+		rte_eth_promiscuous_disable(pd->portid);
+	}
+	rte_eth_dev_stop(pd->portid);
+	rte_eth_dev_close(pd->portid);
+	// free pcap_dpdk?
+	pcap_cleanup_live_common(p);
+} 
+
+static int pcap_dpdk_setfilter(pcap_t *p, struct bpf_program *fp)
+{
+	//init bpf for dpdk, only support userspace bfp 
+	struct pcap_dpdk * pd = p->priv;
+	int ret=0;
+	ret = install_bpf_program(p, fp); 
```

Comment:
If you're calling `install_bpf_program()`, and not passing the filter to the DPDK code to do the filtering at a lower level, you're *always* doing userland filtering, so:

1. you should set `p->setfilter_op` to `install_bpf_program`, as other pcap-XXX.c files that do filtering in userland do (see, for example, `pcap-rdmasniff.c`
1. you should eliminate the `filter_in_userland` structure member, and call `pcap_filter()` if `p->fcode.bf_insns` is non-null.

Reply:
OK, the code have been modified.